File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        int[] suffix = new int[n];

        prefix[0] = 1;
        for(int i=1;i<n;i++){
            prefix[i] = prefix[i-1]*nums[i-1];
        }
        System.out.println(Arrays.toString(prefix));
        // suffix[n-1] = 1;
        // for(int i=n-2;i>=0;i--){
        //     suffix[i] = suffix[i+1]*nums[i+1];
        // }

        // int[] ans = new int[n];
        // for(int i=0;i<n;i++){
        //     ans[i] = prefix[i]*suffix[i];
        // }

        int right = 1;
        for(int i=n-2;i>=0;i--){
            prefix[i] = right*nums[i+1]*prefix[i];
            right*=nums[i+1];
        }

        return prefix;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        larr = [1]
        rarr = [1]
        for i,v in enumerate(nums):
            if i==0:
                continue
            
            larr.append(larr[i-1] * nums[i-1])
        # print(nums)
        nums.reverse()
        # print(nums)
        for i,v in enumerate(nums):
            if i==0:
                continue
            
            rarr.append(rarr[i-1] * nums[i-1])

        rarr.reverse()

        # print(larr)
        # print(rarr)

        ans = [larr[i]*rarr[i] for i in range(0,len(nums))]
        return ans